rfcs/20200205-standalone-keras-repository.md

```diff
@@ -0,0 +1,351 @@
+
+GitHub: the code will live at [keras-team/keras](https://github.com/keras-team/keras), 
+joining the other Keras SIG projects and replacing the current external Keras 
+codebase. `tf.Keras` will also replace Keras on PyPI.
+
+Also considered: `tensorflow/keras`.
+
+| keras-team/keras   | tensorflow/keras |
+:------------------- |:------------------------------------------- |
+|Under the umbrella of Keras SIG, which hosts all other Keras related projects like keras-application, KerasTuner etc.|Under the umbrella of tensorflow, which also hosts other TF related projects.|
+|Lots of existing followers on keras-team, who may not be easily migrated to TF project.|No cross org repo management cost on GitHub. Could rely on a lot of existing setup in TensorFlow.|
+|Can't easily delete keras project, which already have tons of stars and incoming reference links. Continued existence of external Keras code will create confusion ("why is there tensorflow/keras AND keras-team/keras?")|Issue/PR under the same org can be transferred easily, but not cross the different org. See here|
+
+### Source of Truth
+
+TensorFlow uses a Google-internal code repository as its source of truth. Every PR
+submitted though GitHub is converted to a Google-internal change first,
+submitted through the internal system, and then copied to GitHub as commits.
+At the same time, PR is marked as merged with the corresponding commit hash.
+
+Likewise, issue tracking and code review takes place through Google-internal tools.
+
+For Keras, since we are trying to promote community engagement, we hope to use 
+GitHub as source of truth. This will have the following implications:
+
+* We expect the majority of the code development/contribution from GitHub
+and the dev tools / tests / scripts should focus on the GitHub development use
+case. See below for more details.
+* Keras CI/presubmit build for the GitHub repo should target the `tf-nightly` pip
+package as dependency. This means any change to TF will take at most 24
+hours to be reflected on the Keras side.
+* The Keras code will be mirrored to a Google-internal code repository via Google-internal 
+tools within a very short time window after each change.
+The Google-internal CI tests will run on HEAD for both Keras and TF code.
+* The CI build for the repository on GitHub might break when it sees a
+new version of `tf-nightly`, if  certain behavior has been changed and wasn't
+caught by unit tests. We have  observed a few similar cases with
+[tf/addons](https://github.com/tensorflow/addons).
+We hope this can be reduced by stronger
+unit test coverage by Google internel systems, when both TF and Keras code are 
+tested at HEAD.
+* pip package management. Keras will now follow the `tf-estimator` approach. 
+"pip install tensorflow" should also install Keras (from PyPI) as well.
+There are more details for the pip package in the
+[Improved pip package structure](https://github.com/tensorflow/community/pull/182) RFC.
```

Comment:
Can we also consider`conda` this time? `Anaconda` is a huge part of Data Science and ML ecosystem and `tf.keras` not being a part of it doesn't feel right.

Reply:
@gunan from TF build team, and also @seanpmorgan who mentioned this in the SIG-addon meeting.

Reply:
Discussed this item on the design review meeting. So far this is not active target we are supporting. The meeting conclusion are:

Conda uses a different toolchain, with no ABI compatibility between TF+Conda. Because we don’t have C interfaces, any custom ops built for TF will not work for Conda, and vice versa. There is a set of people who publish TF for Conda. But tf addons, io, etc. are not published for Conda. There’s no easy way around this because it would require doubling the TF release burden entirely.
Keras in theory should be easier? Insofar as it is Python only (currently). But Keras will in the future include C(++). That should be fine as long as there is a Python interface to the exposed parts, or C parts are used internally.